File: py/src/schemapb/value_as.py
"""Typed extraction from a wire Value and value path lookup.

``value_as(v, int)`` reads one value as a native Python type, dispatching
on the type object (Python's runtime type token). One rule, shared by
every implementation and pinned by the conformance golden value-as.json: a
conversion succeeds iff the value is represented in the target EXACTLY
(lossless round-trip). Numeric values convert across kinds under that
rule; non-numeric targets are strict - no string parsing, no truncation,
no coercion. Python note: ``int`` is unbounded, so the int token covers
the full int64 AND uint64 wire ranges (the exactness rule with no range
constraint on the target).

``value_lookup`` resolves a path in the ValidationError dialect
("replicas[0].name") against a StructValue - the path from a validation
error fetches the offending value directly. The string form cannot address
a map key containing '.' or '['; the ``value_field``/``value_index``
steppers cover arbitrary keys without parsing.
"""

from __future__ import annotations

import datetime as dt
import json
from enum import StrEnum
from typing import TYPE_CHECKING, overload

import betterproto2

from schemapb.descriptor import join_path
from schemapb.value import struct_v

if TYPE_CHECKING:
    from schemapb._gen.schemapb import StructValue, Value

__all__ = [
    "ValueLookupError",
    "ValueLookupReason",
    "value_as",
    "value_field",
    "value_index",
    "value_kind_name",
    "value_lookup",
]

_KIND_NAMES = {
    "null_value": "null",
    "bool_value": "bool",
    "int32_value": "int32",
    "int64_value": "int64",
    "uint32_value": "uint32",
    "uint64_value": "uint64",
    "float_value": "float",
    "double_value": "double",
    "string_value": "string",
    "bytes_value": "bytes",
    "duration_value": "duration",
    "timestamp_value": "timestamp",
    "list_value": "list",
    "struct_value": "struct",
}

_INT_KINDS = {"int32_value", "int64_value", "uint32_value", "uint64_value"}
_FLOAT_KINDS = {"float_value", "double_value"}


def value_kind_name(v: Value) -> str:
    """The wire kind of a value as its spec short name."""
    name, _ = betterproto2.which_one_of(v, "kind")
    return _KIND_NAMES.get(name, "null" if name == "" else "")


def _value_int(v: Value) -> int | None:
    """The exact integer view of a numeric value (unbounded target)."""
    name, val = betterproto2.which_one_of(v, "kind")
    if name in _INT_KINDS and isinstance(val, int):
        return val
    if name in _FLOAT_KINDS and isinstance(val, float):
        return int(val) if val.is_integer() else None
    return None


def _value_double(v: Value) -> float | None:
    """The exact float64 view of a numeric value."""
    name, val = betterproto2.which_one_of(v, "kind")
    if name in _FLOAT_KINDS and isinstance(val, float):
        return val
    if name in _INT_KINDS and isinstance(val, int):
        f = float(val)
        return f if int(f) == val else None
    return None


@overload
def value_as(v: Value, target: type[bool]) -> bool | None: ...
@overload
def value_as(v: Value, target: type[int]) -> int | None: ...
@overload
def value_as(v: Value, target: type[float]) -> float | None: ...
@overload
def value_as(v: Value, target: type[str]) -> str | None: ...
@overload
def value_as(v: Value, target: type[bytes]) -> bytes | None: ...
@overload
def value_as(v: Value, target: type[dt.timedelta]) -> dt.timedelta | None: ...
@overload
def value_as(v: Value, target: type[dt.datetime]) -> dt.datetime | None: ...
@overload
def value_as(v: Value, target: type[list]) -> list[Value] | None: ...  # type: ignore[type-arg]
@overload
def value_as(v: Value, target: type[dict]) -> dict[str, Value] | None: ...  # type: ignore[type-arg]
def value_as(v: Value, target: type) -> object | None:  # noqa: PLR0911 - flat dispatch per token
    """Read the value as the target type, ``None`` unless exact.

    ``bool`` must come before ``int`` checks: Python bools are ints.
    """
    name, val = betterproto2.which_one_of(v, "kind")
    if target is bool:
        return val if name == "bool_value" else None
    if target is int:
        return _value_int(v)
    if target is float:
        return _value_double(v)
    if target is str:
        return val if name == "string_value" else None
    if target is bytes:
        return val if name == "bytes_value" else None
    if target is dt.timedelta:
        return val if name == "duration_value" else None
    if target is dt.datetime:
        return val if name == "timestamp_value" else None
    if target is list:
        lv = v.list_value
        return list(lv.items) if name == "list_value" and lv is not None else None
    if target is dict:
        sv = v.struct_value
        return dict(sv.fields) if name == "struct_value" and sv is not None else None
    return None


class ValueLookupReason(StrEnum):
    """Stable spec strings shared by all implementations (conformance)."""

    EMPTY_PATH = "empty_path"
    BAD_PATH = "bad_path"
    NOT_FOUND = "not_found"
    INDEX_OUT_OF_RANGE = "index_out_of_range"
    NOT_A_STRUCT = "not_a_struct"
    NOT_A_LIST = "not_a_list"


class ValueLookupError(Exception):
    """Pinpoints the failing segment of a value path.

    ``at`` is the resolved parent path ("" for root), ``segment`` the key
    or "[i]" index that failed.
    """

    def __init__(self, reason: ValueLookupReason, at: str = "", segment: str = "") -> None:
        self.at = at
        self.segment = segment
        self.reason = reason
        super().__init__(self._message())

    def _message(self) -> str:
        where = json.dumps(self.at) if self.at else "root"
        seg = json.dumps(self.segment)
        if self.reason is ValueLookupReason.EMPTY_PATH:
            return "schemapb: value lookup: empty path"
        if self.reason is ValueLookupReason.BAD_PATH:
            return f"schemapb: value lookup: malformed path {seg}"
        if self.reason is ValueLookupReason.NOT_FOUND:
            return f"schemapb: value lookup: no field {seg} in {where}"
        if self.reason is ValueLookupReason.INDEX_OUT_OF_RANGE:
            return f"schemapb: value lookup: index {self.segment} out of range in {where}"
        if self.reason is ValueLookupReason.NOT_A_STRUCT:
            return f"schemapb: value lookup: {where} is not a struct, cannot read field {seg}"
        return f"schemapb: value lookup: {where} is not a list, cannot index {self.segment}"


def value_field(v: Value, name: str) -> Value | None:
    """Step into a struct value member (handles keys paths cannot spell)."""
    sv = v.struct_value
    return sv.fields.get(name) if sv is not None else None


def value_index(v: Value, i: int) -> Value | None:
    """Step into a list value element (``None`` when absent)."""
    lv = v.list_value
    if lv is None or i < 0 or i >= len(lv.items):
        return None
    return lv.items[i]


def _parse_value_path(path: str) -> list[str | int] | None:  # noqa: C901 - one branch per token form
    """Tokenize the error-path dialect: key ('.' key | '[' int ']')*."""
    tokens: list[str | int] = []
    rest = path
    while rest:
        if rest[0] == "[":
            if not tokens:
                return None  # paths start with a key, not an index
            end = rest.find("]")
            body = rest[1:end] if end >= 0 else ""
            # digits only: no "+3", "-0", "[]"
            if end < 2 or not body.isdigit():
                return None
            tokens.append(int(body))
            rest = rest[end + 1 :]
            if rest and rest[0] not in ".[":
                return None  # after "]" only ".", "[" or the end
            continue
        if rest[0] == ".":
            if not tokens:
                return None  # leading dot
            rest = rest[1:]
            # A dot must be followed by a key: no trailing dot, "a..b", "a.[0]".
            if not rest or rest[0] in ".[":
                return None
            continue
        cut = len(rest)
        for i, ch in enumerate(rest):
            if ch in ".[":
                cut = i
                break
        tokens.append(rest[:cut])
        rest = rest[cut:]
    return tokens or None


def value_lookup(s: StructValue, path: str) -> Value:
    """Resolve an error-path-dialect path against the struct's values.

    Returns the addressed value or raises ``ValueLookupError`` naming the
    exact segment that failed.
    """
    if path == "":
        raise ValueLookupError(ValueLookupReason.EMPTY_PATH)

    tokens = _parse_value_path(path)
    if tokens is None:
        raise ValueLookupError(ValueLookupReason.BAD_PATH, "", path)

    cur = struct_v(dict(s.fields))
    parent = ""
    for tok in tokens:
        if isinstance(tok, str):
            sv = cur.struct_value
            if sv is None:
                raise ValueLookupError(ValueLookupReason.NOT_A_STRUCT, parent, tok)
            nxt = sv.fields.get(tok)
            if nxt is None:
                raise ValueLookupError(ValueLookupReason.NOT_FOUND, parent, tok)
            cur = nxt
            parent = join_path(parent, tok)
            continue

        seg = f"[{tok}]"
        lv = cur.list_value
        if lv is None:
            raise ValueLookupError(ValueLookupReason.NOT_A_LIST, parent, seg)
        if tok >= len(lv.items):
            raise ValueLookupError(ValueLookupReason.INDEX_OUT_OF_RANGE, parent, seg)
        cur = lv.items[tok]
        parent += seg

    return cur

File: rust/src/value_as.rs
//! Typed extraction from a wire Value and value path lookup.
//!
//! Extraction is the std-trait way: every target implements
//! `TryFrom<&Value>` (a `Result` with a typed kind-mismatch error), and
//! `Value::get::<T>()` is the `Option` sugar on top. One rule, shared by
//! every implementation and pinned by the conformance golden
//! value-as.json: a conversion succeeds iff the value is represented in
//! the target EXACTLY (lossless round-trip). Numeric values convert
//! across kinds under that rule; non-numeric targets are strict — no
//! string parsing, no truncation, no coercion.
//!
//! `StructValue::lookup` resolves a path in the `ValidationError` dialect
//! ("replicas[0].name") — the path from a validation error fetches the
//! offending value directly. The string form cannot address a map key
//! containing '.' or '['; the `field`/`index` steppers cover arbitrary
//! keys without parsing.

use std::collections::HashMap;

use crate::descriptor::join_path;
use crate::gen::schemapb::value::Kind;
use crate::gen::schemapb::{StructValue, Value};

// =============================================================================
// Kind names and numeric views
// =============================================================================

/// The wire kind of a value as its spec short name.
#[must_use]
pub const fn value_kind_name(v: &Value) -> &'static str {
    match v.kind.as_ref() {
        None | Some(Kind::NullValue(_)) => "null",
        Some(Kind::BoolValue(_)) => "bool",
        Some(Kind::Int32Value(_)) => "int32",
        Some(Kind::Int64Value(_)) => "int64",
        Some(Kind::Uint32Value(_)) => "uint32",
        Some(Kind::Uint64Value(_)) => "uint64",
        Some(Kind::FloatValue(_)) => "float",
        Some(Kind::DoubleValue(_)) => "double",
        Some(Kind::StringValue(_)) => "string",
        Some(Kind::BytesValue(_)) => "bytes",
        Some(Kind::DurationValue(_)) => "duration",
        Some(Kind::TimestampValue(_)) => "timestamp",
        Some(Kind::ListValue(_)) => "list",
        Some(Kind::StructValue(_)) => "struct",
    }
}

#[allow(
    clippy::cast_precision_loss,
    clippy::cast_possible_truncation,
    clippy::float_cmp
)]
fn float_to_i64(f: f64) -> Option<i64> {
    if f.fract() != 0.0 || f < i64::MIN as f64 || f >= i64::MAX as f64 {
        return None;
    }

    let n = f as i64;
    (n as f64 == f).then_some(n)
}

#[allow(
    clippy::cast_precision_loss,
    clippy::cast_possible_truncation,
    clippy::cast_sign_loss,
    clippy::float_cmp
)]
fn float_to_u64(f: f64) -> Option<u64> {
    if f.fract() != 0.0 || f < 0.0 || f >= u64::MAX as f64 {
        return None;
    }

    let n = f as u64;
    (n as f64 == f).then_some(n)
}

/// The exact int64 view of a numeric value.
fn value_int(v: &Value) -> Option<i64> {
    match v.kind.as_ref()? {
        Kind::Int32Value(n) => Some(i64::from(*n)),
        Kind::Int64Value(n) => Some(*n),
        Kind::Uint32Value(n) => Some(i64::from(*n)),
        Kind::Uint64Value(n) => i64::try_from(*n).ok(),
        Kind::FloatValue(f) => float_to_i64(f64::from(*f)),
        Kind::DoubleValue(f) => float_to_i64(*f),
        _ => None,
    }
}

/// The exact uint64 view of a numeric value.
fn value_uint(v: &Value) -> Option<u64> {
    match v.kind.as_ref()? {
        Kind::Int32Value(n) => u64::try_from(*n).ok(),
        Kind::Int64Value(n) => u64::try_from(*n).ok(),
        Kind::Uint32Value(n) => Some(u64::from(*n)),
        Kind::Uint64Value(n) => Some(*n),
        Kind::FloatValue(f) => float_to_u64(f64::from(*f)),
        Kind::DoubleValue(f) => float_to_u64(*f),
        _ => None,
    }
}

/// The exact float64 view of a numeric value.
#[allow(
    clippy::cast_precision_loss,
    clippy::cast_possible_truncation,
    clippy::cast_sign_loss
)]
fn value_double(v: &Value) -> Option<f64> {
    match v.kind.as_ref()? {
        Kind::Int32Value(n) => Some(f64::from(*n)),
        Kind::Uint32Value(n) => Some(f64::from(*n)),
        Kind::Int64Value(n) => {
            let f = *n as f64;
            (f >= i64::MIN as f64 && f < i64::MAX as f64 && f as i64 == *n).then_some(f)
        }
        Kind::Uint64Value(n) => {
            let f = *n as f64;
            (f < u64::MAX as f64 && f as u64 == *n).then_some(f)
        }
        Kind::FloatValue(f) => Some(f64::from(*f)),
        Kind::DoubleValue(f) => Some(*f),
        _ => None,
    }
}

// =============================================================================
// TryFrom<&Value> for every target + Value::get sugar
// =============================================================================

/// A refused extraction: the value's wire kind is not represented in the
/// requested target exactly.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TryFromValueError {
    pub want: &'static str,
    pub got: &'static str,
}

impl std::fmt::Display for TryFromValueError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "schemapb: value is not representable as {} (kind {})",
            self.want, self.got
        )
    }
}

impl std::error::Error for TryFromValueError {}

const fn refused(want: &'static str, v: &Value) -> TryFromValueError {
    TryFromValueError {
        want,
        got: value_kind_name(v),
    }
}

macro_rules! try_from_value {
    ($ty:ty, $want:literal, $v:ident => $extract:expr) => {
        impl TryFrom<&Value> for $ty {
            type Error = TryFromValueError;

            fn try_from($v: &Value) -> Result<Self, Self::Error> {
                $extract.ok_or_else(|| refused($want, $v))
            }
        }
    };
}

try_from_value!(bool, "bool", v => match v.kind.as_ref() {
    Some(Kind::BoolValue(b)) => Some(*b),
    _ => None,
});
try_from_value!(i32, "int32", v => value_int(v).and_then(|n| Self::try_from(n).ok()));
try_from_value!(i64, "int64", v => value_int(v));
try_from_value!(u32, "uint32", v => value_uint(v).and_then(|n| Self::try_from(n).ok()));
try_from_value!(u64, "uint64", v => value_uint(v));
try_from_value!(f32, "float", v => {
    #[allow(clippy::cast_possible_truncation, clippy::float_cmp)]
    value_double(v).and_then(|f| {
        let g = f as Self;
        (f64::from(g) == f).then_some(g)
    })
});
try_from_value!(f64, "double", v => value_double(v));
try_from_value!(String, "string", v => match v.kind.as_ref() {
    Some(Kind::StringValue(s)) => Some(s.clone()),
    _ => None,
});
try_from_value!(Vec<u8>, "bytes", v => match v.kind.as_ref() {
    Some(Kind::BytesValue(b)) => Some(b.clone()),
    _ => None,
});
try_from_value!(pbjson_types::Duration, "duration", v => match v.kind.as_ref() {
    Some(Kind::DurationValue(d)) => Some(*d),
    _ => None,
});
try_from_value!(pbjson_types::Timestamp, "timestamp", v => match v.kind.as_ref() {
    Some(Kind::TimestampValue(t)) => Some(*t),
    _ => None,
});

/// Borrowed string extraction (no allocation).
impl<'a> TryFrom<&'a Value> for &'a str {
    type Error = TryFromValueError;

    fn try_from(v: &'a Value) -> Result<Self, Self::Error> {
        match v.kind.as_ref() {
            Some(Kind::StringValue(s)) => Ok(s.as_str()),
            _ => Err(refused("string", v)),
        }
    }
}

/// Borrowed bytes extraction (no allocation).
impl<'a> TryFrom<&'a Value> for &'a [u8] {
    type Error = TryFromValueError;

    fn try_from(v: &'a Value) -> Result<Self, Self::Error> {
        match v.kind.as_ref() {
            Some(Kind::BytesValue(b)) => Ok(b.as_slice()),
            _ => Err(refused("bytes", v)),
        }
    }
}

impl Value {
    /// `Option` sugar over the `TryFrom<&Value>` impls:
    /// `v.get::<i64>()`, `v.get::<&str>()`, …
    #[must_use]
    pub fn get<'a, T>(&'a self) -> Option<T>
    where
        T: TryFrom<&'a Self>,
    {
        T::try_from(self).ok()
    }

    /// Reads the value as a list (own kind only).
    #[must_use]
    #[allow(clippy::missing_const_for_fn)] // Vec deref is not const-stable
    pub fn as_list(&self) -> Option<&[Self]> {
        match self.kind.as_ref() {
            Some(Kind::ListValue(l)) => Some(&l.items),
            _ => None,
        }
    }

    /// Reads the value as a struct (own kind only).
    #[must_use]
    pub const fn as_struct(&self) -> Option<&HashMap<String, Self>> {
        match self.kind.as_ref() {
            Some(Kind::StructValue(s)) => Some(&s.fields),
            _ => None,
        }
    }

    /// Steps into a struct value member; `None` when the value is not a
    /// struct or has no such field. Handles keys the string path cannot
    /// spell.
    #[must_use]
    pub fn field(&self, name: &str) -> Option<&Self> {
        self.as_struct()?.get(name)
    }

    /// Steps into a list value element; `None` when the value is not a
    /// list or the index is out of range.
    #[must_use]
    pub fn index(&self, i: usize) -> Option<&Self> {
        self.as_list()?.get(i)
    }
}

// =============================================================================
// Value path lookup (the error-path dialect)
// =============================================================================

/// Why a value path failed to resolve. The wire strings are stable spec
/// values shared by all implementations (conformance).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ValueLookupReason {
    EmptyPath,
    BadPath,
    NotFound,
    IndexOutOfRange,
    NotAStruct,
    NotAList,
}

impl ValueLookupReason {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::EmptyPath => "empty_path",
            Self::BadPath => "bad_path",
            Self::NotFound => "not_found",
            Self::IndexOutOfRange => "index_out_of_range",
            Self::NotAStruct => "not_a_struct",
            Self::NotAList => "not_a_list",
        }
    }
}

/// Pinpoints the failing segment of a value path: `at` is the resolved
/// parent path (empty for root), `segment` the key or `[i]` index that
/// failed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ValueLookupError {
    pub at: String,
    pub segment: String,
    pub reason: ValueLookupReason,
}

impl std::fmt::Display for ValueLookupError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let where_ = if self.at.is_empty() {
            "root".to_owned()
        } else {
            format!("{:?}", self.at)
        };

        match self.reason {
            ValueLookupReason::EmptyPath => write!(f, "schemapb: value lookup: empty path"),
            ValueLookupReason::BadPath => {
                write!(
                    f,
                    "schemapb: value lookup: malformed path {:?}",
                    self.segment
                )
            }
            ValueLookupReason::NotFound => {
                write!(
                    f,
                    "schemapb: value lookup: no field {:?} in {where_}",
                    self.segment
                )
            }
            ValueLookupReason::IndexOutOfRange => write!(
                f,
                "schemapb: value lookup: index {} out of range in {where_}",
                self.segment
            ),
            ValueLookupReason::NotAStruct => write!(
                f,
                "schemapb: value lookup: {where_} is not a struct, cannot read field {:?}",
                self.segment
            ),
            ValueLookupReason::NotAList => write!(
                f,
                "schemapb: value lookup: {where_} is not a list, cannot index {}",
                self.segment
            ),
        }
    }
}

impl std::error::Error for ValueLookupError {}

fn verr(reason: ValueLookupReason, at: &str, segment: &str) -> ValueLookupError {
    ValueLookupError {
        at: at.to_owned(),
        segment: segment.to_owned(),
        reason,
    }
}

enum PathToken<'a> {
    Key(&'a str),
    Index(usize),
}

/// Tokenizes the error-path dialect: `key ('.' key | '[' int ']')*`.
fn parse_value_path(path: &str) -> Option<Vec<PathToken<'_>>> {
    let mut tokens = Vec::new();
    let mut rest = path;

    while !rest.is_empty() {
        if let Some(after) = rest.strip_prefix('[') {
            if tokens.is_empty() {
                return None; // paths start with a key, not an index
            }

            let end = after.find(']')?;
            let body = &after[..end];
            // digits only: no "+3", "-0", "[]"
            if body.is_empty() || !body.bytes().all(|b| b.is_ascii_digit()) {
                return None;
            }

            tokens.push(PathToken::Index(body.parse().ok()?));
            rest = &after[end + 1..];

            // After "]" only ".", "[" or the end may follow.
            if !rest.is_empty() && !rest.starts_with('.') && !rest.starts_with('[') {
                return None;
            }

            continue;
        }

        if let Some(after) = rest.strip_prefix('.') {
            if tokens.is_empty() {
                return None; // leading dot
            }

            rest = after;

            // A dot must be followed by a key: no trailing dot, "a..b",
            // "a.[0]".
            if rest.is_empty() || rest.starts_with('.') || rest.starts_with('[') {
                return None;
            }

            continue;
        }

        let end = rest.find(['.', '[']).unwrap_or(rest.len());
        tokens.push(PathToken::Key(&rest[..end]));
        rest = &rest[end..];
    }

    if tokens.is_empty() {
        return None;
    }

    Some(tokens)
}

impl StructValue {
    /// Resolves a path in the `ValidationError` dialect against the struct's
    /// values. Returns the addressed value, or a `ValueLookupError` naming
    /// the exact segment that failed.
    pub fn lookup(&self, path: &str) -> Result<&Value, ValueLookupError> {
        if path.is_empty() {
            return Err(verr(ValueLookupReason::EmptyPath, "", ""));
        }

        let Some(tokens) = parse_value_path(path) else {
            return Err(verr(ValueLookupReason::BadPath, "", path));
        };

        let mut cur: Option<&Value> = None;
        let mut parent = String::new();

        for tok in tokens {
            match tok {
                PathToken::Key(key) => {
                    let fields = match cur {
                        None => &self.fields,
                        Some(v) => v
                            .as_struct()
                            .ok_or_else(|| verr(ValueLookupReason::NotAStruct, &parent, key))?,
                    };

                    cur = Some(
                        fields
                            .get(key)
                            .ok_or_else(|| verr(ValueLookupReason::NotFound, &parent, key))?,
                    );
                    parent = join_path(&parent, key);
                }
                PathToken::Index(i) => {
                    let seg = format!("[{i}]");
                    let cur_v = cur.expect("index token is never first");
                    let items = cur_v
                        .as_list()
                        .ok_or_else(|| verr(ValueLookupReason::NotAList, &parent, &seg))?;

                    cur =
                        Some(items.get(i).ok_or_else(|| {
                            verr(ValueLookupReason::IndexOutOfRange, &parent, &seg)
                        })?);
                    parent.push_str(&seg);
                }
            }
        }

        Ok(cur.expect("tokens are never empty"))
    }

    /// A top-level member of the struct (`None` when absent).
    #[must_use]
    pub fn field(&self, name: &str) -> Option<&Value> {
        self.fields.get(name)
    }
}
